ethereum/eth/src/main/java/org/hyperledger/besu/ethereum/eth/manager/snap/SnapServer.java

```diff
@@ -14,66 +14,616 @@
+                    }
+                  }
+
+                  if (!statefulPredicate.shouldGetMore()) {
+                    break;
+                  }
+                }
+
+                var resp = StorageRangeMessage.create(collectedStorages, proofNodes);
+                LOGGER.debug(
+                    "returned in {} storage {} to {} range {} to {} with {} storages and {} proofs, resp size {} of max {}",
+                    stopWatch,
+                    asLogHash(range.hashes().first()),
+                    asLogHash(range.hashes().last()),
+                    asLogHash(range.startKeyHash()),
+                    asLogHash(range.endKeyHash()),
+                    collectedStorages.size(),
+                    proofNodes.size(),
+                    resp.getSize(),
+                    maxResponseBytes);
+                return resp;
+              })
+          .orElseGet(
+              () -> {
+                LOGGER.debug("returned empty storage range due to missing worldstate");
+                return EMPTY_STORAGE_RANGE;
+              });
+    } catch (Exception ex) {
+      LOGGER.error("Unexpected exception serving storage range request", ex);
+      return EMPTY_STORAGE_RANGE;
+    }
+  }
+
+  MessageData constructGetBytecodesResponse(final MessageData message) {
+    if (!isStarted.get()) {
+      return EMPTY_BYTE_CODES_MESSAGE;
+    }
+    StopWatch stopWatch = StopWatch.createStarted();
+
+    final GetByteCodesMessage getByteCodesMessage = GetByteCodesMessage.readFrom(message);
+    final GetByteCodesMessage.CodeHashes codeHashes = getByteCodesMessage.codeHashes(true);
+    final int maxResponseBytes = Math.min(codeHashes.responseBytes().intValue(), MAX_RESPONSE_SIZE);
+    LOGGER
+        .atTrace()
+        .setMessage("Receive get bytecodes message for {} hashes")
+        .addArgument(codeHashes.hashes()::size)
+        .log();
+
+    try {
+      List<Bytes> codeBytes = new ArrayDeque<>();
+      for (Bytes32 codeHash : codeHashes.hashes()) {
+        Optional<Bytes> optCode = worldStateStorageCoordinator.getCode(Hash.wrap(codeHash), null);
+        if (optCode.isPresent()) {
+          if (sumListBytes(codeBytes) + optCode.get().size() > maxResponseBytes
+              || stopWatch.getTime() > StatefulPredicate.MAX_MILLIS_PER_REQUEST) {
+            break;
+          }
+          codeBytes.add(optCode.get());
+        }
+      }
+      var resp = ByteCodesMessage.create(codeBytes);
+      LOGGER.debug(
+          "returned in {} code bytes message with {} entries, resp size {} of max {}",
+          stopWatch,
+          codeBytes.size(),
+          resp.getSize(),
+          maxResponseBytes);
+      return resp;
+    } catch (Exception ex) {
+      LOGGER.error("Unexpected exception serving bytecodes request", ex);
+      return EMPTY_BYTE_CODES_MESSAGE;
+    }
+  }
+
+  MessageData constructGetTrieNodesResponse(final MessageData message) {
```

Comment:
I think we need to limite the number of trie nodes to serve by request as Geth. Because it seems to be a possible dos vector

Reply:
I will 👀 what geth is doing here.  Just to clarify, what about trie nodes queries makes you think they are disproportionately expensive?  I would think range requests would represent more resource consumption, especially considering we are not directly using an iterator but rather collecting in memory to serve the requests.

Reply:
I ask you 30000 trie nodes that does not exist. You will still have to read in base to find out if the trie node exists. So I can spam you as much as I want. The Range is not the same I ask you for a range and the peers will control what it returns in this range according to the number of leaves. To test it is an intuition that I have